Dev skyrim patcher update

Sharlikran opened the pull request:
This should have the changes from my experimental branch.  I can't guarantee it's 100% the same because of the refactoring and name space changes.  Most of what I have been able to learn comes from working on the Fallout Patchers.  

In my experimental branch I have enabled the same patchers for Fallout as I have for Skyrim.  Irrespective of any differences in opinion I don't really want to keep working on two separate versions.  I would like to add the Fallout support.  It really shouldn't hurt and the Fallout additions are no less functional then with Skyrim.  The Fallout code is very close to Skyrim, a lot closer then Oblivion.  I think it would benefit development of Skyrim patchers.

This is the [first thread](http://forums.bethsoft.com/topic/1507093-wipz-wrye-bashexp-skyrim-patcher-update-thread-1/) for all the testing that was done on the experimental code and has all that has been discussed so far.

Sharlikran commented:
@Utumo I have kept this rebased off of dev, I rebased it again this morning.

Utumno commented:
And now you can see that the HEAD commits in https://github.com/wrye-bash/wrye-bash/commits/dev-skyrim-patcher-update and this branch are identical ;)

I just squashed a bit the commits in mine

Utumno commented:
@Sharlikran I fixed the installers bug and rebased on utumno-163-wip which will be merged to dev soon - so you can test your changes on 306 to be: https://github.com/wrye-bash/wrye-bash/commits/dev-skyrim-patcher-update

Sharlikran commented:
Okay, let me know when you are done so I am not rebasing and testing prematurely.

Utumno commented:
No need to rebase yourself - just checkout your rebased branch by me: https://github.com/wrye-bash/wrye-bash/commits/dev-skyrim-patcher-update and test this

You discovered couple regressions - I guess there are more - keep 'em coming

Feedback on the skyrim patchers is also welcome

Sharlikran commented:
```Auto
Wrye Bash starting
Using Wrye Bash Version 306
Python version: 2.7.8
wxPython version: 2.8.12.1 (msw-unicode)
input encoding: None; output encoding: None; locale: ('en_US', 'cp1252')
Searching for game to manage:
bush.pyo   81 detectGames: Detected the following supported games via Windows Registry:
bush.pyo   83 detectGames:  oblivion: C:\Games\steamapps\common\Oblivion
bush.pyo   83 detectGames:  skyrim: C:\Games\steamapps\common\Skyrim
bush.pyo   95 detectGames: Detecting games via relative path and the -o argument:
bush.pyo  101 detectGames:  skyrim: C:\Games\SteamApps\common\Skyrim
bush.pyo  144 setGame: No preferred game specified.
bush.pyo  151 setGame:  Using skyrim game: C:\Games\SteamApps\common\Skyrim
testing UAC
bosh.pyo 5315 __init__: Using libbsa API version: 2.0.0
bosh.pyo 5321 __init__: Using libloadorder API version: 4.0.1
bosh.pyo 5327 __init__: Using LOOT API version: 0.6.0
C:\Games\SteamApps\common\Skyrim
bush.pyo   81 detectGames: Detected the following supported games via Windows Registry:
bush.pyo   83 detectGames:  oblivion: C:\Games\steamapps\common\Oblivion
bush.pyo   83 detectGames:  skyrim: C:\Games\steamapps\common\Skyrim
bush.pyo   95 detectGames: Detecting games via relative path and the -o argument:
bush.pyo  101 detectGames:  skyrim: C:\Games\SteamApps\common\Skyrim
```

That's all i get so far from that branch.  I was able to get a contextual menu on the mods tab.  I was able to use the Installers tab.  I copied two plugins from the Data folder to the package folder I was using.  One updated the other didn't until I did a refresh of the Data folder.  Which is not a big concern until you do more refactoring.  I'll test it more later but for now it's working better.  If you can get your Skyrim Copy installed, the plugins cleaned, and install all the Unofficial Patches see what the patcher does for you with everything enabled.

Utumno commented:
Thanks!

>  I copied two plugins from the Data folder to the package folder I was using. One updated the other didn't until I did a refresh of the Data folder.

What do you mean the package folder ? Is this new or was present before ?

Sharlikran commented:
It's always been there, [it's a folder rather then an archive](http://sharlikran.github.io/wrye-bash.github.io/Wrye%20Bash%20General%20Readme.html#bain-order).  It looks like a diamond rather the a square.

Utumno commented:
Thanks - I know what a package is - I mean what is the bug exactly ?

Sharlikran commented:
> One updated the other didn't until I did a refresh of the Data folder. Which is not a big concern until you do more refactoring.

It just didn't refresh, but as I mentioned it's not a concern.  I'm not trying to report a bug I'm just observing.

Sharlikran commented:
In regards to the development of Skyrim's patchers I will probably have to wait for other volunteers that have time to work on Wrye Bash with me.  Currently I need people to make the code changes I can't make.  An example of that is, anything that has to do with the back end code.  I can't say whether or not that's 90% of Wrye Bash or 95%, or 99.999%.   Basically I can't make any code changes except updating the record definitions and assigning records, subrecords, or attributes to the patchers.

Since I already explained everything in the previous walls of text then I'll just wait until someone contacts me on STEP, Bethesda, or the Nexus and has time to work on things.  At this point I don't know if I will do anything with this branch until there are volunteers to make the code changes needed.  If that means that this does not make it into 306, 307, 308, 309, 310 then the Bethesda modding community will just have to keep waiting.  I can't program in Python so I can't make the changes needed.

I will make one more pull request with an entirely non working Race Record as an example of what i need help with.  Thanks for your hard work and keep up the refactoring.  Even on STEP some people say they would try to help but the code is too unreadable.

Utumno commented:
Thanks - I understand - when I say I want help I only mean I want you to test the patchers and report what works or not. I do not want to merge non working stuff.

Also having 2 branches for skyrim patchers is a nono. If you test everything in your dev skyrim valda branch then how would I know the one here is ok, So please slowly start rebasing your changes on dev as you 've done here - it's a very good start. I would hope that the tip of this pull request has all the dev valda branch changes.

keep in ming that rebasing can introduce bugs if conflicts resolution is not right - as it happened with the skyrim bsa thing: https://github.com/wrye-bash/wrye-bash/commit/144b707a74c6bba504efff9b1a1164e3ac783a44

Sharlikran commented:
I will double check but I am at least 98% sure I am done integrating all the changes from dev-sharlikran-valda-fallout based off of 305.  I will check this evening but I highly doubt that there is much left to add.  This branch is rebased off of dev the same as dev-fallout-patcher-update.  No rebasing off of any of my older test branches is needed.

Utumno commented:
Great.

If all is done then consider setting up two factor authentication and binning your fork entirely. I believe you now have some git experience so if I could give you write rights here - you can set up your branches and keep working here so we avoid the overhead of the extra branches/pull requests and everything. As long as you do not touch dev pushing (a couple) branches here will make everything much simpler. Plus there are a couple of stale branches here that I wouldn't mind see binned.

Sharlikran commented:
Let me see if I have this right.  

I reowrk things so that I no longer need my fork.  I add TFA, and you give write access.  Just to be clear, you could care less how I manage my branches as long as I make a branch that you can work your magic on and merge into dev.  

Meaning I have a dev-do-smoething branch and it has maybe 100 commits to it.  Then I am done and I make a dev-resolve-somoething branch with less commits.  You look it over, do your fancy git hub gitk squashing rebasing magic, and then merge that rebased/reworked/Utumno-ified into dev.

As long as I do that you are fine?

Utumno commented:
Yep perfectly fine apart maybe from the last part - that is I want you to be eventually able to do the final rebase yourself - I can help you of course but it is you who knows exactly what's going on in your branches. Plus _no flooding of the upstream repo with branches_ (that's a rule for everybody) - have just a couple max triple set up here (in your local repo clone of course you can have whatever you want). Needless to say in those branches you can do whatever you please, that's the deal for `devname-` branches - but if you want them finally reviewed you should clean them up of course - as you said already (dev-resolve-something).

Git `rebase -i` is as essential as git commit, maybe more - I work with it _all the time_ - check its `--onto` flag. And I still recommend msys git gui and gitk - trust me they are brilliant. I always work having all three windows open (git bash, gui, k) - there is a right click menu in gitk that can't be overused (check out branch here, cherry pick, reset branch here etc etc). Plus right clicking on the gitk diff view you can select `show origin of this line` - priceless.

Sharlikran commented:
Then we will have to discuss how to have a Local Repo Clone that I can use with GitHub that is a mirror of wrye-bash but not a Fork.  For example I want to have all the changes to my local repo visible on GitHub and in GFW.  Remember I use 0% Gitk, command line things.  So everything has to be visible on the Web via Chrome/Firefox and GFW.

Sharlikran commented:
> in your local repo clone of course you can have whatever you want

Rember 99.9% of the time I use Chrome/Firefox to view the GitHub webiste and GFW to look at things.  I only open a shell and do what I need to with it 0.01% of the time.  So without the ability to have my own reop that is globally visible to all then I can't make the changes we discussed.  I need to have my own fork of this so I am not violating some rule.

Utumno commented:
You have to start using those tools (or more accurately git) - just give them a whirl - PM me for support. Using the browser to browse your code defeats the purpose of a decentralized system. Plus, I now see that half of your questions are because you can't see the tree locally and because you do not use the gui properly - don't know about GFW but I bet that git gui beats it.

Please install msysgit - guide: http://udunslabs.blogspot.gr/2013/07/git-install-windows.html - you only need first couple paragraphs

Then clone this repo (open a Bash shell by right clicking on a folder you choose and issue `git clone git@github.com:wrye-bash/wrye-bash.git`)

Then right click on the folder and start git gui - from git gui start gitk
![gitkkkkkkkkkkkk](https://cloud.githubusercontent.com/assets/1365079/5750556/f24375e4-9c56-11e4-90b8-0f0c36a88fb0.jpg)

Forget the shell for now - you will need it for the rebases - you'll get there I am sure.

I do understand you are used to work some way but please understand that I have my reasons for asking you. At least try, it's 15 minutes till you see the tree - _you will save me (and you eventually) a bunch of effort and time_.

Excellent pictorial guide: http://nathanj.github.io/gitguide/tour.html

Sharlikran commented:
I have Tortoise git and msysgit installed, I always have had them installed.

![image](https://cloud.githubusercontent.com/assets/1912426/5751697/7fce0310-9c1c-11e4-9632-743aa0183b5a.png)

That does not fix MreRace or get the fallout branch to have a contextual menu when I run it.

Sharlikran commented:
I presented an example of the MreRace record that needs.  To be honest that is not high on my list of priorities right now.  There are probably other things that could be worked on that would be much more simple for you or another volunteer.  Once they are done I could be off on my own updating what I can.  The MreRace record was just the fastest and easiest example I could make.

For now I have just spent the entire day just discussing this and nothing has gotten done.  What I need is to just be able to request help here and there with the code when I hit a snag and can't figure it out.  Sometimes it might mean a few simple changes and other times it might take the majority of the day for someone to make the changes for me that I need.  The GitHub and GitGui part of things really isn't anything I should have to even deal with.  It is for branch maintenance only.  I could have made other changes today if I wasn't worrying and discussing all the GitHub related aspects of this project.  

I have been wanting to work on normalizing the patchers for FO3/FNV/TES4/TES5.  You wanted me to get all my changes based off of dev so that if you did choose to help you are working with what you are familiar with and so that you are working with current code.  So I have done that.  Now the Fallout portion doesn't work.

So now I have to go chill out from all the stress from today.  After that I have to go back and work on my dev-sharlikran-valda-fallout because it compiles and I can run it.

Sharlikran commented:
I don't mind learning to use git enough to properly manage a branch but I feel it is about 5% of what i need to do and I only need to use it 1% of the time.  So I really want to focus as little as possible on GitGui and focus on developing Wrye Bash.

Aside from all the Git stuff this is want I was thinking if I did work from here with write access.

Sharlikran commented:
When there is time to work on Wrye Bash related development of FO3/FNV/TES5 functionality then I will be happy to do so.  For now #185 dev-fallout-patcher-update there is no contextual menu on the mods tab when you right click the plugins.  Removing the lines for the BSA Redirection makes the installers tab work but I don't know what that does for the functionality of INI files for FO3/FNV/TES5.

Sharlikran commented:
> plus all the fixes in dev branch were not included in this rebase here so you were scratching your head in vain

Since you mentioned that in the fallout branch, is that true for this branch?

Utumno commented:
Check gitk ;)

If this branch sprouts from latest dev then it has everything that's in there (plus the fixes) if no, then no

Sharlikran commented:
I used gitk and I get a dif report that looks like this:
![image](https://cloud.githubusercontent.com/assets/1912426/5822949/5a2f350a-a08d-11e4-870b-b5cd7f38f8c3.png)
Then on the GitHub website I used the option to compare the pull request and I get a screen like this
![image](https://cloud.githubusercontent.com/assets/1912426/5822940/478a205e-a08d-11e4-8338-291fef2dc0ba.png)

So while Gitk is useful to see what each individual commit is doing, I find it easier to see a side by side comparison where I know exactly what I am omitting or adding.  Based on what I reviewed today in just the Skyrim branch, it includes all my changes and does not loose things that dev should have that I can see.

If you feel that this branch would force dev to loose your changes, is there a screen shot of what I am loosing from dev, for just the Skyrim branch.

Here is a list of files that I changed in this branch

![image](https://cloud.githubusercontent.com/assets/1912426/5823018/0b1299fc-a08e-11e4-961a-0680b759a4ac.png)

Utumno commented:
Great - this branch is based on dev - so it has the fixes fro mods and installers tab - nothing to do except rebasing on dev soon again :D

Sharlikran commented:
Sweet.  I am trying to follow your guidance and instructions better.  I just hope you understand some views are easier for me to understand then others.

Utumno commented:
Yep I do - gitk is most useful to look the shape of the tree - I sometimes use the online diff of github however (to compare merges).

Keep in ming that gitk has a (very powerful) commit to commit diff

![gitk_compare](https://cloud.githubusercontent.com/assets/1365079/5824941/25b07f54-a0e8-11e4-8f1f-c647787e04fb.jpg)

and options to see inline diffs

![gitk_line_diff](https://cloud.githubusercontent.com/assets/1365079/5824965/4ff05780-a0e8-11e4-9e79-853f5f7fad52.jpg)

but the important is the shape of the tree - the more you'll use it the more you'll see why

Meanwhile I need some feedback on the points I note in last post in #187 - hope you'll get round to it.

Sharlikran commented:
This is up to date with dev now

Sharlikran commented:
The changes in this pull request are much more responsive.  I will continue to see if any of my previous rebasing caused the slow down.  It will still take several days as I have some family related things to attend to.

Sharlikran commented:
@Utumno I think this branch is fine also.  I'll double check this later too.

Arthmoor commented:
Interesting problem, not sure why it happens. Previously, with 306.0.16, A personal fix mod I did for Helgen Reborn would get marked as mergable and show the usual green + in the box.

As of 306.0.17 it is still MARKED as mergable, but when the patch is done building it is getting checkmarked active like a normal mod, even though the name is still green.

In both cases the bashed patch itself merged the contents properly so forcing it to be active is totally redundant. Not sure what's going on. The other mods which can be merged are doing so without a problem.

The file can be downloaded here: http://www.iguanadons.net/HRFixes.7z

Sharlikran commented:
@Arthmoor Go to the beth thread.  I have utumno's version linked now.  `Wrye Bash 306 2015.02.10 02.49` both python and standalone.  If the file you mentioned isn't green anymore, that's because unumno-163-wip doesn't have the Skyrim patcher changes.  Post what you see on the Beth thread for the experimental version.  I have something for you to try is Utumno's version won't let the file be mergable.

Sharlikran commented:
I was able to get the branch rebased off of dev.

Utumno commented:
Excellent

Utumno commented:
@Sharlikran: the time is ripe: I rebased your skyrim changes here: https://github.com/wrye-bash/wrye-bash/compare/sharlikran-skyrim-rebase-on-dev

I do rely on you to help me on merging this.

I would like to merge this in this fashion:
```
- records required for XXX patcher
- XXX patcher
- records required for XXX patcher
- XXX patcher
- ...
```

So I would like you to point me to the first patcher that is ready top be merged and the records it requires. Just comment on the relevant commits mentioning me.

Thanks - and I hope your hand is getting better :)

Sharlikran commented:
Sorry I have not been very active.  My hand injury is very old.  My recent issue was that I had surgery in August but I am fine now of course.  I just don't do much with my computer at the moment.

My opinion has not changed since you asked the last time when I commented on the Forums.  I feel that 306 should not have any Skyrim patchers or new record merging capabilities and should have only your changes.  You will get feedback and bug reports for 306 once it's released.  It will be best if it is only on past bugs people claim have never been resolved or new bugs introduced because of the refactoring.

Target patcher changes for 307.  I also feel patcher changes are best left for 307 because Fallout 4 will be released this November on the 11th.  That is only one month away and people will want to use Wrye Bash with Fallout 4.  All game modes take from the previous games.  meaning that all Fallout 4 records will be Skyrim's records with slight additions and changes.  However, they will be very similar.

The other reason is that as I put in the First Post (the OP or post #1) all the current known issues.  I am sure i have not updated that post with a few things that have been reported.  Those issues will require special unpackers and routines to handle the reocrds that do not currently have routines to handle them.  Like the [OTFT](http://www.uesp.net/wiki/Tes5Mod:Mod_File_Format/OTFT) record can not handle the INAM array correctly.

Utumno commented:
Thanks for feedback - yes I meant 307 for sure. I just want to start merging this branch in dev (or rather the rebased one https://github.com/wrye-bash/wrye-bash/compare/sharlikran-skyrim-rebase-on-dev) which is already 307 - so which records and patchers would be safe to merge - or nothing at all yet ?

Arthmoor commented:
Not sure if anything changed, but the last test version I have with new patchers seems fine. They all work, with the exception of the Stats patcher not handling winning records properly. That's easy to remedy though by removing weapon edits from the patch once it's built.

All of the other ones seem to be functioning properly.

At some point with these, we do need to go over the individual tweaks to evaluate which ones still work for Skyrim. Some don't and may need to be taken out of the files for it.

Utumno commented:
Alright - will include this to my 307 wip ASAP - sharl any objections with my force pushing my rebased branch to this pull request ? Then you pull (on dev say) and force reset your local branch this pull is based on to the upstream (this) one

Sharlikran commented:
Once you are finished with [Making-A-Release](https://github.com/wrye-bash/wrye-bash/wiki/%5Bgithub%5D-Making-A-Release) checklist you can merge it into dev.  The code is fine however, it's the foundation for what needs to be there hence the need for tweaking and updating.

Utumno commented:
Hey sharl - please replace the branch of this pull request with: https://github.com/wrye-bash/wrye-bash/commits/sharlikran-skyrim-rebase-on-dev

Contains all fixes for ordered etc discussed in your thread and also is rebased on recent - does not include loot dll changes

Sharlikran commented:
The thing is that I just redid the entire branch to match dev.  I don't know how many hours it took but possibly 4 or more hours.  I know you have been refactoring for over a year, but still.  As I compare that branch with my newly rebased branch there are too many changes left out of https://github.com/wrye-bash/wrye-bash/commits/sharlikran-skyrim-rebase-on-dev that are in dev as of 491dc027e08de6ae9d311f816ee605fb087289a9.

At this point I don't have a computer that will be able to support Fallout 4 so I won't be able to do anything with that.  Your work is appreciated and I don't mind continuing to work on things a little.  However, since I can't really program in Python in the first place, I can't do anything with Fallout 4, there is little incentive for me to do anything.  So when I do take the time to make changes or updates and then I have to go back and do them another way, I'll probably be less inclined to make the changes in the future.  At this point my computer works great for Playing World of Warcraft with my sons.  So that's probably what I will stick with.

That isn't a rant either.  Nothing you said was offensive or bothered me I'm just being realistic.  I just don't have the hardware or the programming knowledge to be doing this so it's not that rewarding anymore.

Sharlikran commented:
At the moment I was thinking of updating my experimental code to see if with the updates more people are able to use it along with LOOT 0.8.1 and without getting that error that the two users reported on the forum.  If that is the case then the LOOT API update is probably all that was needed to resolve their issue.

Utumno commented:
Please edit out Fallout4 related info - I read what you said in #245 ;)

You did not need to go through rebasing - but still 4 hours is very little - this branch was 400 commits behind. I had done the rebase myself as I said a few posts above.

Anyway now that you did it your self also I rerebased https://github.com/wrye-bash/wrye-bash/commits/sharlikran-skyrim-rebase-on-dev on https://github.com/wrye-bash/wrye-bash/commit/491dc027e08de6ae9d311f816ee605fb087289a9 - so compare this with your rebase (they have now the same base)

Utumno commented:
Now that I reread your message - this is not a rant either but _please read carefully my messages_ - it would have saved you from "make changes or updates and then I have to go back and do them another way" - and would have saved my time too. I posted 16 days ago. What you had to do is have a look at my rebase (then it was on current dev, in 16 days dev moved, of course) and tell me how I can reorganize as I said:

> I would like to merge this in this fashion:.....

I do know you appreciate what I am doing and be aware that I do too - see the credits in the next release ;) (when out).

Utumno commented:
And here you are - the diffs between your rebase and mine:


From cae069903077aae6c0733be6691b94574bf629ad
     Updated class StatsPatcher

To   97dc3422a6f85a7063f35ade9b2ca3b437998776
     Updated class StatsPatcher

```
------------------------------ Mopy/bash/bosh.py ------------------------------
index 4f3bf77..e0f5754 100644
@@ -3381,7 +3381,7 @@ def getStringsPaths(self,language=u'English'):
         if extract:
             bsaPaths = [self.getBsaPath()]
             if bush.game.fsName == u'Skyrim':
-                iniPaths = [ModInfo(self.dir,name).getIniPath() for name in modInfos.ordered]
+                iniPaths = [ModInfo(self.dir,name).getIniPath() for name in modInfos.activeCached]
                 iniFiles = [IniFile(iniPath) for iniPath in iniPaths if iniPath.exists()]
                 iniFiles.reverse()
                 iniFiles.append(oblivionIni)
@@ -4671,7 +4671,7 @@ def isMissingStrings(self,modName):
             sbody,ext = modName.sbody,modName.ext
             bsaPaths = [modInfo.getBsaPath()]
             if bush.game.fsName == u'Skyrim':
-                iniPaths = [ModInfo(self.dir,name).getIniPath() for name in modInfos.ordered]
+                iniPaths = [ModInfo(self.dir,name).getIniPath() for name in modInfos.activeCached]
                 iniFiles = [IniFile(iniPath) for iniPath in iniPaths if iniPath.exists()]
                 # iniFiles.reverse()
                 iniFiles.append(oblivionIni)

----------------------- Mopy/bash/game/skyrim/records.py -----------------------
index 8bb8081..4a4d5b4 100644
@@ -33,8 +33,8 @@
     ModSizeError, MelObject, MelGroups, MelStruct, FID, MelGroup, MelString, \
     MreLeveledListBase, MelSet, MelFid, MelNull, MelOptStruct, MelFids, \
     MreHeaderBase, MelBase, MelUnicode, MelModel, MelFidList, MelStructA, \
-    MreRecord, MreGmstBase, MelLString, MelCountedFidList,  MelOptStructA, \
-    MelCountedFids, MelSortedFidList, ModReader, MelStrings
+    MreRecord, MreGmstBase, MelLString, MelCountedFidList, MelOptStructA, \
+    MelCountedFids, MelSortedFidList, MelStrings, ModReader # why ModReader ?
 from ...bass import null1, null2, null3, null4
 from ... import bush
 from constants import allConditions, fid1Conditions, fid2Conditions, \
@@ -996,7 +996,7 @@ def dumpData(self):
                     else:
                         raise Exception(u'Unrecognized VMAD property type: %s' % type(Type))
             else:
-                raise Exception(u'Unrecognized VMAD property type: %s' % type(Type))
+                raise Exception(u'Unrecognized VMAD property type: %s' % type(value))
             return data
 
     class Script(object):
```

My version is the correct one.

So reset this branch to my version

Because "there is no bad with no good" as the saying goes - now that we both did the rebase and we ended up with nearly identical branches we can be sure the rebased branch is correct.

My rebase however contains squashing of fixes with the commits that introduced - ___it is very important we do not push broken commits on dev___.

Finally in commit: https://github.com/wrye-bash/wrye-bash/commit/50f32fa684bd8c256ee0c9f4b483023f7d71081a the author was accidentally set to me - leave this to me when i merge to dev

So now please let me know which records go with which patchers - so i may merge this patcher after patcher - preferably

Utumno commented:
Let loot out of this for now - one thing at a time

Utumno commented:
I just had a look at your thread - if you need to pack a version with newer loot go ahead - name it 307.0.0 and base it on my rebase - one of the differences is the `ordered` traceback ;)

Sharlikran commented:
> Now that I reread your message - this is not a rant either but please read carefully my messages

I had read them.

> You did not need to go through rebasing - but still 4 hours is very little - this branch was 400 commits behind. I had done the rebase myself as I said a few posts above.

I knew that but once I took a look at it there were differences.

Now.  I have absolutely no idea how to make sharlikran-skyrim-rebase-on-dev become dev-skyrim-patcher-update and overwrite the history.  You have write access so feel free to do it.  However please only replace the dev-skyrim-patcher-update.

Utumno commented:
Thanks for following up - i will delete dev-skyrim-patcher-update from this repo and leave sharlikran-skyrim-rebase-on-dev . You can sync your pull request here by usual means (see also brand new screens in https://github.com/wrye-bash/wrye-bash/wiki/%5Bgit%5D-Syncing-a-Fork-with-the-main-repository)

But tommorowishzzzz

Sharlikran commented:
![new gitk rebase for exp branch](https://cloud.githubusercontent.com/assets/1912426/11166381/d4f64e84-8ae5-11e5-85a0-aa71c14e5e4f.png)

So what I see in GitK is more or less what I expected.  Basically my branch dev-skyrim-pathcer-update is the same as 151-skyrim-patchers, then the loot changes, then the fallout changes.  Is that what KitK is telling me in that screen shot?

Utumno commented:
Yep, you have become a git geek over time :)

Issue an exp release when you come round to it - there should be nothing really new in the patchers but there is in the code - plus we can't get enough testing.

Sharlikran commented:
I was wondering @Utumno does [this wiki](https://github.com/wrye-bash/wrye-bash/wiki/%5Bgit%5D-Syncing-a-Fork-with-the-main-repository) cover history rewrites?  I have one of my branches rebased on your 151-skyrim-patchers so that if I change anything I don't change that branch and you can cherry pick any changes.  I was going to rebase that branch again and instead of fast forwarding to your last change it started rebasing changes.

Utumno commented:
@Sharlikran:

>I have one of my branches rebased on your 151-skyrim-patchers so that if I change anything I don't change that branch and you can cherry pick any changes. 

Good ! (links...)

> I was going to rebase that branch again and instead of fast forwarding to your last change it started rebasing changes.

Expected - have a look at the tree - my branch is rewritten - rebased on dev (and some commits rewritten at that) so your branch can't be fast forwarded - the commits it is based on are no more in the base branch.

Sharlikran commented:
I will work on squashing the commits you mentioned.  However, because of some recent changes in RL I probably won't have time to complete that until Friday or Saturday.

Utumno commented:
Please do - I want to merge non records patchers commits - so at least please squash the https://github.com/Sharlikran/wrye-bash/commit/47a9906682859e670d247266fc05fdd45f23c3d4 one and edit it and let me cherry pick it

Sharlikran commented:
I though I would have more time to do this a few weeks ago.  Due to changes at work I barley had time to get a Christmas tree and presents for my daughters.  I have a few days off and may have to time to work on this but I can't be certain yet.  Sorry for the delay.

Utumno commented:
And a happy new year Sharl :)

No worries - I run in a bug storm and have tricky merges upcoming so I can wait a couple weeks - ideally max. If rebasing is hard I can do it - just point me to a neat branch containing all the needed commits and be ready to answer me questions. Keep in mind that I will be merging ASAP some serious refactoring in bosh that may affect the patchers/records etc. Ideally you could also rebase on my wip branch to see how that goes.

If however you want to push directly this branch here I can now give you write rights. Please be very careful - make sure your password is ok now ! PM me when ready

Sharlikran commented:
In regards to squashing new changes I made to the Skyrim Patchers.  Originally I had a branch named dev-skyrim-patcher-update which you turned into 151-skyrim-patchers.  I then rebased the dev-skyrim-patcher-update on 151-skyrim-patchers and made some minor changes.  You requested that I squash some of the commits which means changing 151-skyrim-patchers.

Not all of them can be squashed because the change didn't occur in 151-skyrim-patchers and instead change commits in dev.  If at some point you merge 151-skyrim-patchers into dev then more squashing would be possible.

Because I still don't feel comfortable altering 151-skyrim-patchers directly I made a copy of the branch named upstream-151-skyrim-patchers.  Feel free to review the changes and then work your Git magic.

Utumno commented:
You mean I should go ahead and rebase upstream-151-skyrim-patchers on dev and squash those myself ?

Sharlikran commented:
Were you planning on merging 151-skyrim-patchers into dev at soon?  I was thinking you might want to compare your 151-skyrim-patchers and upstream-151-skyrim-patchers first and then decide what to do just to be safe.

Utumno commented:
Compared them - so what we need merged is upstream-151-skyrim-patchers right ? will rebase that on dev and rename it to 151-skyrim-patchers (replace that one) and you will have a chance to review it your self - ASAP, that is tommorowish ;)

Sharlikran commented:
I think we are on the same page.  Because upstream-151-skyrim-patchers is your 151-skyrim-patchers plus my https://github.com/Sharlikran/wrye-bash/tree/dev-skyrim-patcher-update in this pull request.  Once you rebase upstream-151-skyrim-patchers on dev and rename it to 151-skyrim-patchers 
```
dev \
     151-skyrim-patchers \
                          Sharlikran fork : dev-skyrim-pathcer-updates
```
Would be equal to :
```
dev \
     upstream-151-skyrim-patchers \
```

Utumno commented:
Now that I had a closer look I realize that you followed my instructions to the letter squashing the commits as I told you up to "minor ini edits" - that's really appreciated. Aparently your git foo is now at a serious level. That's very good news to me and I feel rewarded for all the efforts I put. Thanks and congrats !

This means we can now collaborate efficiently - and god knows Bash needs that.

As promised I pushed the 151-skyrim-patchers branch - and went ahead and deleted the upstream-151-skyrim-patchers, feature-sharlikran-valda-fallout, sharlikran-fallout-skyrim-rebase-on-dev branches from here - for the last two when their time comes we will follow a similar procedure, having an up to date, rebased on dev branch.

Now let's concentrate on 151-skyrim-patchers. As you see I moved to the beginning some commits that are not directly patcher related which I intend to merge first - the 5 first up to ini edits. Any other such commits (as little stuff, corrections etc) now is the time to merge - feel free to push them directly to 151-skyrim-patchers and comment on them so I move them to the beginning. Meanwhile this is the time to rewrite some commit messages to detail what's going on - commits are a sure permanent place to safeguard the experience you have gathered working on the patchers. So rewrite commit messages detailing (if not obvious) what changes they introduce and why - for instance: https://github.com/wrye-bash/wrye-bash/commit/ff16356dd292938fa4d4d2755ded32889b3a3e8d - what is this new class ? (btw this is exactly what I was talking about that you followed my instructions on squashing 5 commits to this one - now what remains to be done is a nice commit message explaining what goes on).

From the first commits I need especially commented https://github.com/wrye-bash/wrye-bash/commit/a0ec758500974ec1d53ff9b3cd335aa8f3d01b39 - especially does this close the relevant issue ?

If you don't feel like rebasing and rewriting the commit messages, adding them as commit comments will also do and will rewrite the messages my self on final rebase merge.

Thanks indeed Sharl :+1:

Sharlikran commented:
> That's very good news to me and I feel rewarded for all the efforts I put. Thanks and congrats !

You are welcome and your assistance has been appreciated.  I may continue to use methods and tools that you wouldn't but, the end result should be close to what you want.

> Aparently your git foo is now at a serious level

I wouldn't place that much confidence in my git foo.  I have learned how to do what I need/want and that's about it.  Not that I am unappreciative of your help.  I simply don't know as much about Git as you do.

> for instance: ff16356 - what is this new class ?

Done see https://github.com/wrye-bash/wrye-bash/commit/7ab3505d8b93a1ec46d0579ce6be96f48d4d1205

> I need especially commented a0ec758 - especially does this close the relevant issue ?

Lojack has two checkboxes in Issue #148.  Commit a0ec758500974ec1d53ff9b3cd335aa8f3d01b39 addresses the second checkbox not the first.  

If it were up to me I would make two issues out of #148.  The new issue would address the first checkbox where Lojack asks the question--in what order does Skyrim process BSA files when assigned by the along side INI file in the Data folder.  We know how BSA files are processed when it's only a plugin and a BSA file but, we probably don't know how BSA files are processed from the along side INI files.  That should be a separate issue and researched when time permits.

Leave the history and comments in #148 alone because it pertains to processing the INI file which @ tony971 had requested and he has already commented on the functionality of the update.  Then a0ec758500974ec1d53ff9b3cd335aa8f3d01b39 would be more directly tied to #148.  We could then ask for more testing to make sure when we can close the issue.

I'd like to request that you make the new issue because I will have no idea how to word it and whatever I made you would edit anyway.

Sharlikran commented:
@Utumno I still don't know exactly how to pull in changes you make.  However, I'm sure you know how to do that for mine.  Be sure to compare my changes in 151-skyrim-patchers with your copy before you pull from the repository.  When I squashed commits and updated comments it rewrote the history and I'd like you to check and make sure I didn't mess something up. :pray:

Utumno commented:
Haha - no worries - just pull this branch in your fork and base an exp release on this.

See now why it is important to rebase early ? Cause anyway at some point one has to and then there are too many changes to review.

Better concentrate on the commits as opposed to the diffs from now on - your changes are all there, I can guarantee that, cause I really did not mess with skyrim records myself at all and I had no conflicts while rebasing. You can do the rebase yourself - I just rearranged some stuff but if you compare the tips of the rebased branches they are identical. What we need to do is merge the non patcher stuff and then carefully review the rest and merge. Writing some commit messages makes one look at the commits more carefully (and it's super boring I know :P)

Sharlikran commented:
So trying to keep my pull request attached to all this history and not knowing git very well has my dev-skyrim-patcher-update attached to multiple branches.

![multibranch](https://cloud.githubusercontent.com/assets/1912426/12373241/cd29f7aa-bc27-11e5-98c4-128300ba73b0.jpg)

Is there a way to clean that up without deleting the branch or disassociating the branch from this pull request?

Utumno commented:
Haha - no keep it it will be an incentive to merge faster :D

Well we could just keep the  dev-skyrim-patcher-update branch and have the rest deleted ? Or maybe just leave it for now. The important thing is that they all point to same content - it really did not help me at all to have this branch storm. One branch to rule them all - and now that you have your write rights no more pull requests - they just add branch noise, working with branches directly is more elegant, less error prone and much more a la git. And everyone who uses the tree can immediately see what goes on.

I see you squashed the tabs fixing commit in its parent right ?

Also I missed your first message - re: #148 I am inclined to leave it as it is and let @lojack5 handle it. Will just merge the second check box and wait for users to ~~start complaining~~ give me feedback

Sharlikran commented:
> it really did not help me at all to have this branch storm

I didn't intend to do that.  Why is that happening?  Notifications?  The dev-skyrim-patcher-update is like my staging ground for anything related to Skyrim or Oblivion, such as taglist.yaml file updates and such.  That way it keeps things separate.

I can just commit all Skyrim and Oblivion changes to 151-skyrim-patchers but I really didn't want to get rid of dev-skyrim-patcher-update until all of it is in dev.  Because most likely I will update something.  For example, after I rebase all my branches if users report a bug or if I update another patcher I would put it in dev-skyrim-patcher-update because you are working with 151-skyrim-patchers.  I added the comments you asked for but I want to limit changes to 151-skyrim-patchers while you decide what to merge into dev.  If I keep changing 151-skyrim-patchers, even if I do it cleanly and neatly it might mess something up for you and I don't want to deal with that.

So storm or not, I want to make minimal changes to branches, and have the least amount of branches on wrye-bash.  That is my goal at least.

> I see you squashed the tabs fixing commit in its parent right ?

Yes.  All those tab errors belonged to 0a02a816ee78f3a1377e5ba880a660125b33ab7d.

Sharlikran commented:
In regards to 151-skyrim-patchers:

![first commits](https://cloud.githubusercontent.com/assets/1912426/12373506/e68a23e8-bc2f-11e5-8d66-198c59f7238a.jpg)

Those are the commits you want to merge.  Is that correct?  I am done commenting or changing those.  Let me know if you intend to merge more commits then what I show in the above screen shot.

Utumno commented:
Alright sharl - but this way whenever you make a change in  dev-skyrim-patcher-update (like the (correct !) squashing you made of the tabs errors) you have to let me know to pull it in 151-skyrim-patchers. So better get rid of 151-skyrim-patchers and all the tracking branches and just keep this  dev-skyrim-patcher-update branch as I don't thing one can change the branch of a pull request. And make a deal: I won't edit (or merge) anything without notifying you and whenever you rewrite commits (be it messages or rebases on dev or squashes or ...) and you push them you let me know (in the commit comments or here, depending on the change) so I am aware and I can pull the updated branch.

This way we are on the same boat - that's what I meant with branch storm, even with two branches it's a huge issue, for instance I was about to squash the commit you squashed ! So yes, let's delete all other branches from here apart this pull and work from here

Utumno commented:
Ooops ninja'ed: yes those are the commits - I want to review a bit the imnportant (as in code changes) one (the 148 inis one) and then I can go ahead and merge. So this may take a couple days - I would rather need @lojack5 opinion on that

Sharlikran commented:
I want to head toward something you feel comfortable with.  I am trying to work with you not against you.  If you prefer to use 151-skyrim-patchers lets stick with only that branch.  I'd rather keep the comments history here and start a new issue for Skyrim's patchers.

Utumno commented:
No no I am perfectly comfortable with the dev pull - they have instructions above on how to pull it in my local tree and I've already done that. So go ahead and delete the 151 branches from your fork

Utumno commented:
Please don't open an new issue anyway - #151 is all we need. Anyway I follow youy here - do as you please just we end up with _one_ branch and if we edit this we notify each other - deal ? :)
